stephenxs/sonic-buildimage
# split initialization of sonic_platform to a per-daemon basis

**- What I did**
split sonic_platform initialization to a per-daemon basis to save the startup time of daemons
make Mellanox new platform api able to be installed on host so that reboot cause can be handled on it.

**- How I did it**
1. modify makefiles so that the following modules are installed on host:
    - sonic_platform, which handles reboot cause
    - sonic_daemon_base, on which sonic_platform depends
2. remove unnecessary dependencies of sonic_platform on host
    - implement platform.py
    - add an added argument to sonic_platform.platform.Platform's constructor, as following:
    `Platform.__init__(self, daemon = None)`
      - [TBD] the base class sonic_platform_base.platform_base.PlatformBase.__init__ should be updated accordingly
    - Platform's constructor loads and initializes only the functionalities on a per-daemon basis, i.e. only functionalities required by the daemon are initialized for it.
    - define initializers for each components belonging the chassis and remove them out of the chassis's constructor so that they can be called only when needed. currently the following initializers defined:
      - initialize_psu
      - initialize_watchdog
      - initialize_fan
      - initialize_sfp
      - initialize_eeprom
      - initialize_components_list

**- How to verify it**
load chassis on host and check whether reboot-cause and watchdog can be well handled.
load chassis on pmon docker and check whether other components can be well handled.
run process-reboot-cause and check whether the correct reboot cause can be parsed.


**- Description for the changelog**



**- A picture of a cute animal (not mandatory but encouraged)**
